### DISCUSSION TITLE
Loudnorm/replay gain correction on the fly

### DISCUSSION THREAD

**Discussion by iSerganov** [savonet/liquidsoap#4231] [Q&A] [answered]
Hey team!

Currently I am using Liquidsoap 2.2.5 + autocue 4.4.1 (https://github.com/Moonbase59/autocue) to compute track loudness normalization, cue in/out points together with fading in & out.

That's how I configure it:
```python
# Ref: https://github.com/Moonbase59/autocue?tab=readme-ov-file#parameters-and-settingsreference
settings.autocue.cue_file.noclip := true  
settings.autocue.cue_file.nice := true  
settings.autocue.cue_file.fade_in := fade_in
settings.autocue.cue_file.fade_out := fade_out
settings.autocue.cue_file.timeout := cue_exec_timeout
settings.autocue.cue_file.silence := cue_silence_threshold
settings.autocue.cue_file.overlay := cue_overlay
settings.autocue.cue_file.sustained_loudness_drop := sustained_loudness_drop
settings.autocue.cue_file.target := cue_target
```

Liquidsoap 2.3.0 documentation says:
```
Autocue, the method for computing the perfect crossfade has also been polished in this release. We have now switched to our internal implementation, which is available when enabling our ffmpeg support.

You should be able to enable it by adding this to your script:

enable_autocue_metadata()

With this, all requests should resolve with the appropriate metadata so that any crossfade produces perfect transitions!
```

So I have 2 questions:

1. Can you please confirm that`autocue` can be completely replaced by some Liqudisoap internal mechanism doing the same things?
2. Where can I find a complete guide for overall loudness level control and loudness normalization/replay gain correction? Maybe I was doing something wrong, but steps describen in this page https://www.liquidsoap.info/doc-dev/replay_gain.html do not work with Liquidsoap 2.2.5, that's why [autocue](https://github.com/Moonbase59/autocue) was really my savior.

Thank you in advance!

**Comment by gAlleb** [comment]
Hi @iSerganov 

You can read how `autocue` started in this small presentation (https://moonbase59.github.io/autocue/presentation/autocue.html)

There are 2 autocue implementations:

1. Moonbase59 autocue with cue_file which you have been using (e.g. I'm using it too with LS 2.3.0). It uses external python script with ffmpeg (and mutagen sometimes).
2. Internal LS autocue. It uses ffmpeg inside LS.


They have similar functionality and settings. Which one to use is up to you. Moonbase59 stopped development of his cue_file autocue due to personal reasons a few month ago. LS internal autocue has been evolving and as of now provides you with good functionality (exception: long lossless tracks). Current cue_file implementation is not 100% tuned for use with LS 2.3.0 but still provides good results.


> Can you please confirm thatautocue can be completely replaced by some Liqudisoap internal mechanism doing the same things?

Liqudisoap internal mechanism is called autocue and it's designed in a way that you can use it as it is or attach to it any other autocue implementation. You can see it's settings right in the library file: https://github.com/savonet/liquidsoap/blob/main/src/libs/autocue.liq
However it lacks `no-clip` option and could be slower sometimes. E.g. LUFS target option for internal autocue: `settings.autocue.internal.lufs_target := -13.0` - will do exactly what it states.

All you have to do is:

```
enable_autocue_metadata()
s = playlist("~/playlist")
s = amplify(1., override="liq_amplify", s)
s = crossfade(s)
```
It will take care of the rest, normalizing loudness and making transitions between tracks smooth.

<hr/>

> Where can I find a complete guide for overall loudness level control and loudness normalization/replay gain correction? Maybe I was doing something wrong, but steps describen in this page https://www.liquidsoap.info/doc-dev/replay_gain.html do not work with Liquidsoap 2.2.5, that's why [autocue](https://github.com/Moonbase59/autocue) was really my savior.

They should. I don't have 2.2.5 at hand, but in 2.3.0 they work. All you need is:

```
enable_replaygain_metadata()

s = playlist("~/playlist")
s = amplify(1., override="replaygain_track_gain", s)
```

or

```
enable_replaygain_metadata()

s = replaygain(playlist("~/playlist"))
```

or

```
s = replaygain(playlist(prefix="replaygain:","~/playlist"))
```

or

```
s = playlist(prefix="replaygain:","~/playlist")
s = amplify(1., override="replaygain_track_gain", s)
```

The above 4 examples do the same thing. They compute `replaygain_track_gain` (if it's not present in the tags) and apply its value via `amplify` operator.

Both autocues (internal and cue_file) also normalize loudness similar to replaygain mechanism but they use ffmpeg EBU R128 scanner filter. Plus cue_file allows no-clipping option.

The best option for loudness normalization is of course pre-tagging your files using good software that can do it (e.g. `loudgain` - https://github.com/Moonbase59/loudgain). Then you won't spend any CPU on it and can use just this:

```
# No prefixes, no extra functions.
s = playlist("~/playlist")
s = amplify(1., override="replaygain_track_gain", s)
```

But this won't give you auto `cue_in/cue_out` or auto `fade_in/out`.

**Comment by iSerganov** [comment]
Thank you for the detailed explanation, @gAlleb!
After playing around for some time with internal autocue I have figured out how to make smooth transitions and loundess normalization with its help.

However still have 2 open questions:

1. `cue_file` provides much more metadata per each track after on the fly computation (I think Python script does metadata tagging there). Whereas with internal `autocue` I don't have even track's duration available per a file. Is there a way to extend the metadata tag set available as a result of `autocue` work? E.g. the same track duration I guess is something you compute within `autocue` for internal purposes, why not to make it available in `source.on_metadata` callback?
3. Is there a way to disable `cue_in/cue_out` and auto `fade_in/out` for tracks under a certain duration?
It looks like `settings.autocue.internal.sustained_endings_min_duration` should be doing this according to the comment below:
`Minimum duration to consider it the ending as sustained [seconds]`, however it is not 100% clear for me what kind of duration is meant here.
Could you please confirm or deny that if I set `settings.autocue.internal.sustained_endings_min_duration := 60`, I will completely disable `fade_in/out` for all tracks with duration under 60s (other tracks won't be affected)?

**Reply by gAlleb** [reply]
> why not to make it available in source.on_metadata callback?

1. All is in your hands. You can make your own function. E.g. something like what AC uses:

```
def log_meta(m) = 
    label="log_meta"
    l = list.sort.natural(metadata.cover.remove(m))
    list.iter(fun(v) -> log(level=4, label=label, "#{v}"), l)

    nowplaying = ref(m["artist"] ^ " - " ^ m["title"])

    if m["artist"] == "" then
        if string.contains(substring=" - ", m["title"]) then
            let (a, t) = string.split.first(separator=" - ", m["title"])
            nowplaying := a ^ " - " ^ t
        end
    end

    # show `liq_` & other metadata in level 3
    def fl(k, _) =
        tags = ["duration", "media_id", "replaygain_track_gain", "replaygain_reference_loudness"]
        string.contains(prefix="liq_", k) or list.mem(k, tags)
    end

    liq = list.assoc.filter((fl), l)
    list.iter(fun(v) -> log(level=3, label=label, "#{v}"), liq)
    log(level=3, label=label, "Now playing: #{nowplaying()}")
end

radio = source.on_metadata(radio, log_meta)
```

> Is there a way to disable cue_in/cue_out and auto fade_in/out for tracks under a certain duration?
It looks like settings.autocue.internal.sustained_endings_min_duration should be doing this according to the comment below:
Minimum duration to consider it the ending as sustained [seconds], however it is not 100% clear for me what kind of duration is meant here.
Could you please confirm or deny that if I set settings.autocue.internal.sustained_endings_min_duration := 60, I will completely disable fade_in/out for all tracks with duration under 60s (other tracks won't be affected)?


2. Sustained endings - is when a song has a long sustained ending (not a hard end) and you decide what you wanna do with it. If you want to disable autocue for a specific track you can annotate it with something like `annotate:liq_disable_autocue=true:` or with manual cue/fade values.

**Reply by iSerganov** [reply]
Maybe I am doing something wrong, however the output metadata is exactly the same in both cases after calling
`metadata.cover.remove(m)` and just using `m`. There are no `liq_` prefixed tags or `duration` or anything besides tags from the annotations and from original file metadata.


```
enable_autocue_metadata()
# Log dir
log.file.path.set("/tmp/basic-radio.log")

myplaylist = playlist("/playlist.pls",reload_mode="watch",mode="normal")
radio = amplify(1., override="liq_amplify", myplaylist)

def on_track_change(m) =
  full_meta = list.sort.natural(metadata.cover.remove(m))
 
  def fl(k, _) =
    tags = ["duration", "media_id", "replaygain_track_gain", "replaygain_reference_loudness"]
    string.contains(prefix="liq_", k) or list.mem(k, tags)
  end

  liq = list.assoc.filter((fl), full_meta)
  list.iter(fun(v) -> log(level=1, label="log metadata", "#{v}"), liq)
end

radio = source.on_metadata(radio, on_track_change)

```

**Reply by iSerganov** [reply]
Or I am missing something fundamental... 
Probably `cue_file` does certain pre-tagging therefore those tags were available, whereas internal `autocue` do not set all of those tags?

**Reply by iSerganov** [reply]
>     2. Sustained endings - is when a song has a long sustained ending (not a hard end) and you decide what you wanna do with it. If you want to disable autocue for a specific track you can annotate it with something like `annotate:liq_disable_autocue=true:` or with manual cue/fade values.

If I set any of these, it will completely exclude the file from processing according to:
```
let settings.autocue.internal.metadata_override =
  settings.make(
    description=
      "Disable processing when one of these metadata is found",
    [
      "liq_cue_in",
      "liq_cue_out",
      "liq_fade_in",
      "liq_fade_in_delay",
      "liq_fade_out",
      "liq_fade_out_delay",
      "liq_disable_autocue"
    ]
  )
```
However I still need loudness normalization for the tracks where I want to disable fading-in/out. Is there a way to achieve this?

**Reply by gAlleb** [reply, marked as answer]
> Maybe I am doing something wrong, however the output metadata is exactly the same in both cases after calling
metadata.cover.remove(m) and just using m. There are no liq_ prefixed tags or duration or anything besides tags from the annotations and from original file metadata.

Everything works fine with me. You don't get duration tag because internal autocue doesn't provide it, butt all liq_ value are intact (file in my playlist doesn't initially have liq_ tags). Full script:

```
settings.log.stdout := true
log.file.path.set("./basic-radio.log")
enable_autocue_metadata()
myplaylist = playlist("./music/1.m3u")

radio = amplify(1., override="liq_amplify", myplaylist)

def on_track_change(m) =
  full_meta = list.sort.natural(metadata.cover.remove(m))
 
  def fl(k, _) =
    tags = ["duration", "media_id", "replaygain_track_gain", "replaygain_reference_loudness"]
    string.contains(prefix="liq_", k) or list.mem(k, tags)
  end

  liq = list.assoc.filter((fl), full_meta)
  list.iter(fun(v) -> log(level=1, label="log metadata", "#{v}"), liq)
end

radio = source.on_metadata(radio, on_track_change)
radio = crossfade(radio)
output.dummy(mksafe(radio))
```

```
2024/12/04 08:39:08 [cross.buffer:3] Content type is {audio=pcm(stereo)}.
2024/12/04 08:39:08 [log metadata:1] ("liq_amplify", "-7.832 dB")
2024/12/04 08:39:08 [log metadata:1] ("liq_autocue", "internal")
2024/12/04 08:39:08 [log metadata:1] ("liq_cross_end_duration", "1.5")
2024/12/04 08:39:08 [log metadata:1] ("liq_cross_max_start_duration", "181.5")
2024/12/04 08:39:08 [log metadata:1] ("liq_cross_start_duration", "0.0")
2024/12/04 08:39:08 [log metadata:1] ("liq_cue_in", "0.0")
2024/12/04 08:39:08 [log metadata:1] ("liq_cue_out", "183.0")
2024/12/04 08:39:08 [log metadata:1] ("liq_fade_in", "0.0")
2024/12/04 08:39:08 [log metadata:1] ("liq_fade_out", "1.5")
2024/12/04 08:39:08 [log metadata:1] ("liq_fade_out_delay", "0.0")
2024/12/04 08:39:08 [log metadata:1] ("liq_fade_out_start_next", "0.0")
2024/12/04 08:39:08 [ao:3] Opening MacOS X AUHAL output (2 channels)...
```

<hr/>

If you need duration for some reason you can alter the function a bit, e.g.:

```
def on_track_change(m) =
  file_duration = null.get(default=0.0, request.duration(m["filename"]))

  full_meta = list.sort.natural(metadata.cover.remove(m))
 
  def fl(k, _) =
    tags = ["duration", "media_id", "replaygain_track_gain", "replaygain_reference_loudness"]
    string.contains(prefix="liq_", k) or list.mem(k, tags)
  end

  liq = list.assoc.filter((fl), full_meta)
  list.iter(fun(v) -> log(level=1, label="log metadata", "#{v}"), liq)
  log(level=1, label="log metadata", "file_duration: #{file_duration}")

end
```

Then you'll get

```
2024/12/04 08:51:11 [autocue.internal.sustained_ending:3] No sustained ending detected.
2024/12/04 08:51:11 [log metadata:1] ("liq_amplify", "-7.832 dB")
2024/12/04 08:51:11 [log metadata:1] ("liq_autocue", "internal")
2024/12/04 08:51:11 [log metadata:1] ("liq_cross_end_duration", "1.5")
2024/12/04 08:51:11 [log metadata:1] ("liq_cross_max_start_duration", "181.5")
2024/12/04 08:51:11 [log metadata:1] ("liq_cross_start_duration", "0.0")
2024/12/04 08:51:11 [log metadata:1] ("liq_cue_in", "0.0")
2024/12/04 08:51:11 [log metadata:1] ("liq_cue_out", "183.0")
2024/12/04 08:51:11 [log metadata:1] ("liq_fade_in", "0.0")
2024/12/04 08:51:11 [log metadata:1] ("liq_fade_out", "1.5")
2024/12/04 08:51:11 [log metadata:1] ("liq_fade_out_delay", "0.0")
2024/12/04 08:51:11 [log metadata:1] ("liq_fade_out_start_next", "0.0")
2024/12/04 08:51:11 [log metadata:1] file_duration: 186.685
```
But I don't recommend it, cause requesting duration may cause pauses sometimes.

<hr/>

>However I still need loudness normalization for the tracks where I want to disable fading-in/out. Is there a way to achieve this?

You can achieve this with a combination of settings:

```
settings.autocue.internal.metadata_override := ["liq_disable_autocue"]
settings.autocue.metadata.priority := 1
```

1. First setting tells LS not to disable autocue if we have `liq_fade_in` and so on metadata and annotations. And to disable it ONLY if we have `liq_disable_autocue`.
2. "Priority for the autocue metadata resolver. Default value (`10`) allows it to override both file and request metadata." We set it to `1` and now it doesn't override your manual annotations but at the same time keeps computing LUFS target:

```
annotate:liq_cue_in=90.0,liq_fade_in="0.0",liq_fade_out="0.0":/music/track.flac
```

```
2024/12/04 09:52:15 [log metadata:1] ("liq_amplify", "-7.832 dB")
2024/12/04 09:52:15 [log metadata:1] ("liq_autocue", "internal")
2024/12/04 09:52:15 [log metadata:1] ("liq_cross_end_duration", "1.5")
2024/12/04 09:52:15 [log metadata:1] ("liq_cross_max_start_duration", "181.5")
2024/12/04 09:52:15 [log metadata:1] ("liq_cross_start_duration", "0.0")

2024/12/04 09:52:15 [log metadata:1] ("liq_cue_in", "90.0")

2024/12/04 09:52:15 [log metadata:1] ("liq_cue_out", "183.0")

2024/12/04 09:52:15 [log metadata:1] ("liq_fade_in", "0.0")
2024/12/04 09:52:15 [log metadata:1] ("liq_fade_out", "0.0")

2024/12/04 09:52:15 [log metadata:1] ("liq_fade_out_delay", "0.0")
2024/12/04 09:52:15 [log metadata:1] ("liq_fade_out_start_next", "0.0")

```

This way it will keep `liq_fade_in` and `liq_fade_out` but will compute everything else.

Also beware that default target is `-14.0` LUFS. It can be changed with `settings.autocue.internal.lufs_target`

**Reply by iSerganov** [reply]
Ah, now I know why I got pauses - I retrieve duration exactly as you advised. 🤣 
OK, now it seems clear to me. Thank you - I thinks that answers all my questions!
